cherab/solps/models/line_emitter.pyx

```diff
@@ -0,0 +1,111 @@
+# CONDITIONS OF ANY KIND, either express or implied.
+#
+# See the Licence for the specific language governing permissions and limitations
+# under the Licence.
+
+from raysect.optical cimport Spectrum, Point3D, Vector3D
+from cherab.core cimport Plasma, AtomicData
+from cherab.core.model.lineshape cimport GaussianLine, LineShapeModel
+from cherab.core.utility.constants cimport RECIP_4_PI
+from cherab.core.math.mappers import AxisymmetricMapper
+
+from cherab.solps cimport SOLPSFunction2D
+
+
+cdef class SOLPSLineEmitter(PlasmaModel):
+    """
+    Line emitter that uses a custom `SOLPSFunction2D` function for radiation power density.
+
+    :param SOLPSFunction2D radiation_function: The power density function defined on the SOLPS mesh.
+    :param Line line: Spectral line.
+    :param Plasma plasma: Plasma object. Default is None.
+    :param AtomicData atomic_data: Atomic data provider. Will use `OpenADAS` if None (default).
+    :param object lineshape: Spectral line shape model. Will use `GaussianLine` if None (default).
+    :param object lineshape_args: Positional arguments for the lineshape.
+    :param object lineshape_kwargs: Keyword arguments for the lineshape.
+    """
+
+    def __init__(self, SOLPSFunction2D radiation_function, Line line, Plasma plasma=None, AtomicData atomic_data=None,
+                 object lineshape=None, object lineshape_args=None, object lineshape_kwargs=None):
+
+        super().__init__(plasma, atomic_data)
+
+        self._line = line
+
+        self._power_density = AxisymmetricMapper(radiation_function)
+
+        self._lineshape_class = lineshape or GaussianLine
+        if not issubclass(self._lineshape_class, LineShapeModel):
+            raise TypeError("The attribute lineshape must be a subclass of LineShapeModel.")
+
+        if lineshape_args:
+            self._lineshape_args = lineshape_args
+        else:
+            self._lineshape_args = []
+        if lineshape_kwargs:
+            self._lineshape_kwargs = lineshape_kwargs
+        else:
+            self._lineshape_kwargs = {}
+
+        # ensure that cache is initialised
+        self._change()
+
+    def __repr__(self):
+        return '<SOLPSLineEmitter: element={}, charge={}, transition={}>'.format(self._line.element.name, self._line.charge, self._line.transition)
+
+    cpdef Spectrum emission(self, Point3D point, Vector3D direction, Spectrum spectrum):
+
+        cdef double ne, ni, te, radiance
+
+        # cache data on first run
+        if self._lineshape is None:
+            self._populate_cache()
+
+        # add emission line to spectrum
+        radiance = RECIP_4_PI * self._power_density(point.x, point.y, point.z)
+        return self._lineshape.add_line(radiance, point, direction, spectrum)
+
+    cdef int _populate_cache(self) except -1:
+
+        # sanity checks
+        if self._plasma is None or self._atomic_data is None:
+            raise RuntimeError("The emission model is not connected to a plasma object.")
```

Comment:
Separate this into two checks. Otherwise if a user manages to supply a plasma but no atomic data they will get a misleading error message.

Reply:
I fixed this and also checked other plasma models from Cherab/core. The `ExcitationLine`, `RecombinationLine`, `TotalRadiatedPower` and `BeamCXLine` models are affected by the same issue. I'll rise the issue and prepare a fix in Cherab/core.